src/core/api.rs

```diff
@@ -174,168 +460,157 @@ impl Covercrypt {
 }
 
-/// Encrypted header holding a `Covercrypt` encapsulation of a symmetric key and
-/// additional data encrypted using the `Covercrypt` DEM with the encapsulated
-/// key.
-///
-/// *Note*: the DEM ciphertext is also used to select the correct symmetric key
-/// from the decapsulation.
-///
-/// - `encapsulation`       :   `Covercrypt` encapsulation of a symmetric key
-/// - `encrypted_metadata`  :   AES-256 GCM encryption of the metadata
-#[derive(Debug, PartialEq, Eq)]
-pub struct EncryptedHeader {
-    pub encapsulation: Encapsulation,
-    pub encrypted_metadata: Option<Vec<u8>>,
-}
-
-impl EncryptedHeader {
-    /// Generates an encrypted header for a random key and the given metadata.
-    /// Returns the encrypted header along with the symmetric key
-    /// encapsulated in this header.
+impl<
+        const KEY_LENGTH: usize,
+        const NONCE_LENGTH: usize,
+        const MAC_LENGTH: usize,
+        E: AE<KEY_LENGTH, NONCE_LENGTH, MAC_LENGTH>,
+    > CovercryptPKE<E, KEY_LENGTH, NONCE_LENGTH, MAC_LENGTH> for Covercrypt
+{
+    /// Encrypts the given plaintext using Covercrypt and the given DEM.
     ///
-    /// - `cover_crypt`         : `Covercrypt` object
-    /// - `policy`              : global policy
-    /// - `public_key`          : `Covercrypt` public key
-    /// - `encryption_policy`   : access policy used for the encapsulation
-    /// - `header_metadata`     : additional data symmetrically encrypted in the
-    ///   header
-    /// - `authentication_data` : authentication data used in the DEM encryption
-    pub fn generate(
-        cover_crypt: &Covercrypt,
+    /// Creates a Covercrypt encapsulation of a LENGTH-byte key, and use this
+    /// key with the given authentication data to produce a DEM ciphertext of
+    /// the plaintext.
+    ///
+    /// # Error
+    ///
+    /// Returns an error if the access policy is not valid.
+    fn encrypt(
+        &self,
+        mpk: &MasterPublicKey,
         policy: &Policy,
-        public_key: &MasterPublicKey,
-        encryption_policy: &AccessPolicy,
-        metadata: Option<&[u8]>,
-        authentication_data: Option<&[u8]>,
-    ) -> Result<(SymmetricKey<SEED_LENGTH>, Self), Error> {
-        let (symmetric_key, encapsulation) =
-            cover_crypt.encaps(policy, public_key, encryption_policy.clone())?;
-
-        let encrypted_metadata = metadata
-            .map(|bytes| cover_crypt.dem_encrypt(&symmetric_key, bytes, authentication_data))
-            .transpose()?;
-
-        Ok((
-            symmetric_key,
-            Self {
-                encapsulation,
-                encrypted_metadata,
-            },
-        ))
+        ap: &str,
+        ad: Option<&[u8]>,
+        plaintext: &[u8],
+    ) -> Result<(Encapsulation, Vec<u8>), Error> {
+        let (seed, enc) = CovercryptKEM::encaps(self, mpk, policy, ap)?;
+        let mut sym_key = Secret::<KEY_LENGTH>::default();
+        kdf256!(&mut sym_key, &seed);
```

Comment:
The hash can be avoided if we truncate. Anyway, there is not enough entropy to create a key of greater length:
```suggestion
        if SEED_LENGTH < Self::KEY_LENGTH {
            return Err(...);
        }
        let key = SymmetricKey::try_from_slice(&seed[..Self::KEY_LENGTH])?;
```